enhancements/20251015-support-ipconfig-on-clusternetworks.md

```diff
@@ -0,0 +1,416 @@
+
+#### Story 12: Create VM Network with cluster network `mgmt`, vid `2014`,`L3 Connectivity` field set and mode `DHCP`
+
+- bridge vlan vid `2014` added to `mgmt` on node 1,node and node 3
+- vlan sub interface `mgmt-br.2014` is created on node 1,node 2,node 3 in the cluster and ip address is assigned from DHCP.
+
+### User Experience In Detail
+
+User will be able to create vlan interfaces on required nodes in the cluster and assign IP address to them using static or DHCP mode and will be able to persist this across reboots.
+This could be achieved either from UI by providing the network config details on the VM Network page or updating the network config info on VM Network annotations manually.
+
+### API changes
+
+None.
+
+## Design
+
+### Implementation Overview
+
+#### Overview
+
+- Support the following user configuration on the annotations of the Virtual Machine Network resource
+  - IPconnectivity Required(true/false)
+  - Mode(Static/DHCP)
+  - Mode (Static) - List of Node to IP address mapping 
+
+case 1: VM Network Resource with mode Static
+
+```
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  annotations:
+    network.harvesterhci.io/route: >-
+      {"mode":"auto","cidr":"10.115.8.0/21","gateway":"10.115.15.254","connectivity":"true"}
+    network.harvesterhci.io/l3-connectivity: "true"
+    network.harvesterhci.io/mode: "static"
+    network.harvesterhci.io/node-ip-map: |
+      {
+        "node1": "192.168.1.10/24",
+        "node2": "192.168.1.11/24",
+        "node3": "192.168.1.12/24"
+      }
+  labels:
+    network.harvesterhci.io/clusternetwork: cluster-1
+    network.harvesterhci.io/ready: 'true'
+    network.harvesterhci.io/type: L2VlanNetwork
+    network.harvesterhci.io/vlan-id: '2012'
+  name: vlan2012
+  namespace: default
+spec:
+  config: >-
+    {"cniVersion":"0.3.1","name":"vlan2012","type":"bridge","bridge":"cluster-1-br","promiscMode":true,"vlan":2012,"ipam":{}}
+```
+case 2: VM Network Resource with mode DHCP
+
+```
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  annotations:
+    network.harvesterhci.io/route: >-
+      {"mode":"auto","cidr":"10.115.8.0/21","gateway":"10.115.15.254","connectivity":"true"}
+    network.harvesterhci.io/l3-connectivity: "true"
+    network.harvesterhci.io/mode: "dhcp"
+  labels:
+    network.harvesterhci.io/clusternetwork: cluster-1
+    network.harvesterhci.io/ready: 'true'
+    network.harvesterhci.io/type: L2VlanNetwork
+    network.harvesterhci.io/vlan-id: '2012'
+  name: vlan2012
+  namespace: default
+spec:
+  config: >-
+    {"cniVersion":"0.3.1","name":"vlan2012","type":"bridge","bridge":"cluster-1-br","promiscMode":true,"vlan":2012,"ipam":{}}
+```
+
+- Harvester network controller handles the nad change and updates the annotations of the cluster network with the following.
+  - VLAN-ID (2 - 4094)
+  - Mode (Static or DHCP)
+  - List of nodenames to IP address mapping (if Mode = Static)
```

Comment:
Based on my understanding, is it correct that the only thing that matters is the ClusterNetwork but not NAD? It looks like we chose NAD as the entry point for this feature because it contains VLAN information. So technically, the user can bypass updating the NAD and instead annotate the ClusterNetwork with sufficient information (mode, VLAN ID, etc), and it still works as expected. Please correct if I misunderstood.